### PR TITLE
Save proofs even when successful

### LINKED ISSUE #2183: Save proofs even when successful
`--save-proofs` should save the trusted claims even when the specification is successfully proved.

### PR DESCRIPTION
**Pull request by ttuegel**
Fixes #2183 

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

**Comment by ttuegel** [comment]
This should just about be ready. I only need to add some documentation for types I added.

File: test/save-proofs/test-2-spec.k [hunk 1]
```diff
@@ -10,6 +10,8 @@ endmodule
 module TEST-2-SPEC
   imports VERIFICATION
 
+  // This claim would fail, but it is loaded from the saved claims.
   rule BB(X:Int) => CC(X:Int)
+  // This claim requires the first claim.
```
Comment (ana-pantilie) [inline test/save-proofs/test-2-spec.k:15]: Is this test missing an expected output?

File: test/save-proofs/test-1-spec.k [hunk 2]
```diff
@@ -1,13 +1,12 @@
-// Two claims, one passes and is saved to a file, the other one fails.
 module VERIFICATION
   imports SAVE-PROOFS
 
   rule BB(X:Int) => CC(X:Int)
 endmodule
 
+// This claim is proven and saved.
```
Comment (ana-pantilie) [inline test/save-proofs/test-1-spec.k:7]: Where is it saved?

File: kore/test/Test/Kore/Reachability/Prove.hs [hunk 3]
```diff
@@ -69,308 +61,127 @@ import Test.Tasty.HUnit.Ext
...
+                        Unlimited
+                        (Limit 3)
+                        [ simpleAxiom Mock.a Mock.a ]
+                        [ mkSimpleClaim Mock.a Mock.b ]
+                        []
+                    assertEqual "" MultiAnd.top actual
+        in
+        [ mkTest "OnePath" simpleOnePathClaim
+        , mkTest "AllPath" simpleAllPathClaim
+        ]
```
Comment (ana-pantilie) [inline kore/test/Test/Kore/Reachability/Prove.hs:184]: Any reason why we don't want the mixed cases anymore?

File: kore/test/Test/Kore/Reachability/Prove.hs [hunk 4]
```diff
@@ -402,28 +211,7 @@ test_reachabilityVerification =
...
-            , simpleAxiom Mock.a Mock.c
-            ]
-            [ simpleOnePathClaim Mock.a (mkOr Mock.b Mock.c)
-            , simpleOnePathClaim Mock.a (mkOr Mock.b Mock.c)
-            ]
-            []
-        assertEqual ""
-            (Right ())
-            actual
+        assertEqual "" MultiAnd.top actual
```
Comment (ana-pantilie) [inline kore/test/Test/Kore/Reachability/Prove.hs:214]: Could these tests be combined like the tests above were?

**Review by ana-pantilie — COMMENTED** [review]
I added some questions.

In addition, I saw that you cleaned up the tests in `kore/test/Test/Kore/Reachability/Prove.hs`, but some of them kept the previous structure (I added a comment on one instance of this). Is there any reason for this, or could we (the reviewers) try to clean up those tests as well, as part of the reviewer refactoring?

File: kore/src/Kore/Reachability/Prove.hs [hunk 5]
```diff
@@ -160,31 +165,35 @@ The action may throw an exception if the proof fails; the exception is a single
...
+    isBottom = const False
+
+type StuckClaims = MultiAnd StuckClaim
+
+type ProvenClaims = MultiAnd SomeClaim
+
+-- | The result of proving some claims.
+data ProveClaimsResult =
+    ProveClaimsResult
+    { stuckClaim :: !StuckClaims
```
[the remark this answers is not in the document]
Comment (ttuegel) [reply kore/src/Kore/Reachability/Prove.hs:192]: > Maybe a middle ground solution would be `stuckClaims`, with the check in Main (line 703) renamed to `noStuckClaims` (or something similar) in a where clause.

That sounds good to me!

> Another question, in what situation can we have more than one stuck claim? I thought we stop execution when we find the first one.

When we exceed the breadth limit, we will present all the branches for the user to help with debugging.

File: kore/test/Test/Kore/Reachability/Prove.hs (hunk 3 above)
Comment (ttuegel) [reply kore/test/Test/Kore/Reachability/Prove.hs:184]: The only thing that these mixed cases actually tested is that the backend doesn't try to use one-path claims to prove all-path claims, or vice versa. There are already tests specifically for that, and I didn't think it made sense to duplicate that check too many times.

File: kore/test/Test/Kore/Reachability/Prove.hs (hunk 4 above)
Comment (ttuegel) [reply kore/test/Test/Kore/Reachability/Prove.hs]: Yes, I cleaned up the test suite some more.

**Review by ana-pantilie — APPROVED** [review]

**Review by andreiburdusa — APPROVED** [review]

**Pull request merged**